### PR TITLE
Hide un-needed DSO symbols

### PR DESCRIPTION
**Pull request by ssheorey**
- [x] Hide all symbols except `PyInit_pybind` from Python module in GNU/Linux and macOS.
- [x] Hide all non Open3D symbols in Open3D DLLs in GNU/Linux and macOS.
- [ ] Future PR: Hide these on Windows.
- [x] Use Build Kit in docker, since disabling it causes CI error in ARM64 CI.

### Motivation:
- Reduce open3d DSO / import open3d load time by speeding up symbol resolution.
- Reduce Open3D DSO / python wheel size (slightly).
- Prevent unexplained crashes when Open3D and another library simultaneously load different versions of a library into the same process (e.g. Python or an app using Open3D). (e.g. #2926) This issue should be completely solved after this PR.
- Previous related PR: #3542. In contrast to that PR, this PR also hides 3rd party symbols from header only libraries (eigen, fmt, etc.)
- Related issue #5124 

### Linux

19MB reduction in installed wheel size on Linux.
12MB reduction in installed wheel size on macOS.

On master:
```sh
nm -CD --defined-only cpu/pybind.cpython-38-x86_64-linux-gnu.so | wc -l
28937
nm -CD --defined-only cuda/pybind.cpython-38-x86_64-linux-gnu.so | wc -l
76551
ls -hs cpu/pybind.cpython-38-x86_64-linux-gnu.so
185M
ls -hs cuda/pybind.cpython-38-x86_64-linux-gnu.so
779M
```

This PR:
```
nm -CD --defined-only cpu/pybind.cpython-38-x86_64-linux-gnu.so | wc -l
2
nm -CD --defined-only cuda/pybind.cpython-38-x86_64-linux-gnu.so | wc -l
2
du -h cpu/pybind.cpython-38-x86_64-linux-gnu.so | cut -f1
182M
ls -hs cuda/pybind.cpython-38-x86_64-linux-gnu.so | cut -f1
763M
```

### macOS

On master:
```sh
nm -gU cpu/pybind.cpython-310-darwin.so | c++filt | wc -l
17355
cut -h cpu/pybind.cpython-38-x86_64-linux-gnu.so | cut -f1
127M
```

This PR:
```
nm -gU cpu/pybind.cpython-310-darwin.so | c++filt | wc -l
1
du -h cpu/pybind.cpython-310-darwin.so | cut -f1
115M
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/isl-org/Open3D/5877)

### CODE REVIEW

**Review by benjaminum — COMMENTED** [review]
@ssheorey Are the changes sufficiently covered by the CI? I think we build some example apps but is this sufficient?

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/isl-org/Open3D/5877#-:-NNvP48OCxeu2QDGo5pl:b63givo)* status: all files reviewed (commit messages unreviewed), all discussions resolved (waiting on @yxlao)

File: docker/docker_build.sh [hunk 1]
```diff
@@ -82,545 +80,545 @@ CMAKE_VERSION=cmake-3.20.6-linux-x86_64
 CMAKE_VERSION_AARCH64=cmake-3.20.6-linux-aarch64
 
 print_usage_and_exit_docker_build() {
-    echo "$__usage_docker_build"
-    exit 1
+	echo "$__usage_docker_build"
```
Comment (yxlao) [inline docker/docker_build.sh:83]: Shall we keep using (4) spaces as indentations, as we use in other places?

**Review by ssheorey — COMMENTED** [review]
In the CI with BUILD_SHARED_LIBS=ON, `libOpen3D.{dylib,so}` is linked to `tests` (unit test executable) and all examples. Since the tests run correctly, the hidden symbols aren't needed for Open3D functionality.

> *[Reviewable](https://reviewable.io/reviews/isl-org/Open3D/5877#-:-NNwIAbV1FohkFzb-FWq:b-pfzqtq)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @ssheorey)

File: docker/docker_build.sh (hunk 1 above)
Comment (ssheorey) [reply docker/docker_build.sh:83]: Done

**Review by benjaminum — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/isl-org/Open3D/5877#-:-NOGwqeg87Z2uvoiHpoD:bnfp4nl)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @yxlao)

**Pull request merged**